homeassistant/components/yolink/sensor.py

```diff
@@ -109,14 +113,23 @@ def cvt_battery(val: int | None) -> int | None:
     return 0
 
 
-def cvt_volume(val: int | None) -> str | None:
+def cvt_volume(val: int | None, device: YoLinkDevice) -> str | None:
     """Convert volume to string."""
     if val is None:
         return None
     volume_level = {1: "low", 2: "medium", 3: "high"}
     return volume_level.get(val, None)
 
 
+def cvt_distance(val: int | None, device: YoLinkDevice) -> float | None:
+    """Distance conversion."""
+    if device.device_attrs is not None and val is not None:
+        dev_range = device.device_attrs["range"]["range"]
+        dev_density = device.device_attrs["range"]["density"]
+        return round((dev_range * (val / 1000)) / dev_density, 2)
```

Comment:
We don't allow calculating state for entities that are part of integrations that integrate devices or services. It's ok to convert from one prefix to another prefix, eg from k to M or similar but other calculations that involve other unit of measurements aren't allowed. Time state is the only exception.

The API data should be used as is, while following our entity design guidelines.